provider/oci/provider.go

```diff
@@ -0,0 +1,336 @@
+		return errors.NotSupportedf("%q auth-type", authType)
+	}
+	return nil
+}
+
+// PrepareConfig implements environs.EnvironProvider.
+func (e EnvironProvider) PrepareConfig(args environs.PrepareConfigParams) (*config.Config, error) {
+	if err := validateCloudSpec(args.Cloud); err != nil {
+		return nil, errors.Annotate(err, "validating cloud spec")
+	}
+	// TODO(gsamfira): Set default block storage backend
+	return args.Config, nil
+}
+
+// Open implements environs.EnvironProvider.
+func (e *EnvironProvider) Open(params environs.OpenParams) (environs.Environ, error) {
+	logger.Infof("opening model %q", params.Config.Name())
+
+	if err := validateCloudSpec(params.Cloud); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	creds := params.Cloud.Credential.Attributes()
+	provider := providerCommon.NewJujuConfigProvider(
+		creds["user"], creds["tenancy"],
+		[]byte(creds["key"]), creds["fingerprint"],
+		creds["pass-phrase"], creds["region"])
+
+	client, err := providerCommon.NewOciClient(provider)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	env := &Environ{
+		cli:   client,
+		p:     e,
+		clock: clock.WallClock,
+	}
+
+	if err := env.SetConfig(params.Config); err != nil {
+		return nil, err
+	}
+
+	env.namespace, err = instance.NewNamespace(env.Config().UUID())
+
+	cfg := env.ecfg()
+	if cfg.compartmentID() == nil {
+		return nil, errors.New("compartment-id may not be empty")
+	}
+
+	return env, nil
+}
+
+// CredentialSchemas implements environs.ProviderCredentials.
+func (e EnvironProvider) CredentialSchemas() map[cloud.AuthType]cloud.CredentialSchema {
+	return credentialSchema
+}
+
+func validateKey(key []byte, passphrase string) error {
+	keyBlock, _ := pem.Decode(key)
+	if keyBlock == nil {
+		return errors.Errorf("invalid private key")
+	}
+
+	if x509.IsEncryptedPEMBlock(keyBlock) {
+		if _, err := x509.DecryptPEMBlock(keyBlock, []byte(passphrase)); err != nil {
+			return errors.Annotatef(err, "decrypting private key")
+		}
+	}
+
+	return nil
+}
+
+// DetectCredentials implements environs.ProviderCredentials.
+func (e EnvironProvider) DetectCredentials() (*cloud.CloudCredential, error) {
+	cfg_file, err := ociConfigFile()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	cfg, err := ini.LooseLoad(cfg_file)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	cfg.NameMapper = ini.TitleUnderscore
+
+	result := cloud.CloudCredential{
+		AuthCredentials: make(map[string]cloud.Credential),
+	}
+
+	var defaultRegion string
+
+	for _, val := range cfg.SectionStrings() {
+		values := new(credentialSection)
+		if err := cfg.Section(val).MapTo(values); err != nil {
+			return nil, errors.Annotatef(err, "invalid value in section %s", val)
+		}
+
+		if values.User == "" || values.Tenancy == "" ||
+			values.KeyFile == "" || values.Fingerprint == "" || values.Region == "" {
+			return nil, errors.Errorf("missing required fields in config section %s", val)
```

Comment:
Does it make sense exit the entire loop if one of the sections is invalid?
It seems better to read the sections that are valid and have the user be able to use them.
Probably with some sort of warning about sections that are incomplete.

Especially here, all the user ends up seeing is "missing a required field", but not mention of what fields are actually required, and what fields appear to be missing.

Reply:
No, it does not make sense to exit. Adding detailed warning, and continuing.